aws/resource_aws_servicecatalog_portfolio_principal_association.go

```diff
@@ -0,0 +1,163 @@
+	_, err := conn.AssociatePrincipalWithPortfolio(&input)
+	if err != nil {
+		return fmt.Errorf("creating Service Catalog Principal(%s)/Portfolio(%s) Association failed: %s",
+			principalArn, portfolioId, err.Error())
+	}
+
+	stateConf := resource.StateChangeConf{
+		Pending:      []string{servicecatalog.StatusCreating},
+		Target:       []string{servicecatalog.StatusAvailable},
+		Timeout:      1 * time.Minute,
+		PollInterval: 3 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			err := resourceAwsServiceCatalogPortfolioPrincipalAssociationRead(d, meta)
+			if err != nil {
+				return 42, "", err
+			}
+			if d.Id() != "" {
+				return 42, servicecatalog.StatusAvailable, err
+			}
+			return 0, servicecatalog.StatusCreating, err
+		},
+	}
+	_, err = stateConf.WaitForState()
+	return err
+}
+
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationRead(d *schema.ResourceData, meta interface{}) error {
+	id, portfolioId, principalArn := resourceAwsServiceCatalogPortfolioPrincipalAssociationRequiredParameters(d)
+	input := servicecatalog.ListPrincipalsForPortfolioInput{
+		PortfolioId: aws.String(portfolioId),
+	}
+	conn := meta.(*AWSClient).scconn
+	var pageToken = ""
+	isFound := false
+	for {
+		pageOfDetails, nextPageToken, err := resourceAwsServiceCatalogPortfolioPrincipalAssociationListPrincipalsForPortfolioPage(conn, input, &pageToken)
+		if err != nil {
+			return err
+		}
+		for _, principal := range pageOfDetails {
+			if *principal.PrincipalARN == principalArn {
+				isFound = true
+				d.SetId(id)
+				break
+			}
+		}
+		if nextPageToken == nil || isFound {
+			break
+		}
+		pageToken = *nextPageToken
+	}
+	if !isFound {
+		log.Printf("[WARN] Service Catalog Principal(%s)/Portfolio(%s) Association not found, removing from state",
+			principalArn, portfolioId)
+		d.SetId("")
+	}
+	d.Set("principal_arn", principalArn)
+	d.Set("portfolio_id", portfolioId)
+	return nil
+}
+
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationListPrincipalsForPortfolioPage(conn *servicecatalog.ServiceCatalog, input servicecatalog.ListPrincipalsForPortfolioInput, nextPageToken *string) ([]*servicecatalog.Principal, *string, error) {
+	input.PageToken = nextPageToken
+	var page, err = conn.ListPrincipalsForPortfolio(&input)
+	if err != nil {
+		return nil, nil, fmt.Errorf("retrieving Service Catalog Associations for Principal/Portfolios: %s", err.Error())
+	}
+	principalDetails := page.Principals
+	return principalDetails, page.NextPageToken, nil
+}
+
+func resourceAwsServiceCatalogPortfolioPrincipalAssociationUpdate(d *schema.ResourceData, meta interface{}) error {
+	if d.HasChanges("principal_arn", "portfolio_id") {
+		oldPrincipalArn, newPrincipalArn := d.GetChange("principal_arn")
+		oldPortfolioId, newPortfolioId := d.GetChange("portfolio_id")
+		d.Set("principal_arn", oldPrincipalArn.(string))
+		d.Set("portfolio_id", oldPortfolioId.(string))
+		resourceAwsServiceCatalogPortfolioPrincipalAssociationDelete(d, meta)
```

Comment:
this is a bit unusual as far as i can tell, same thing for the call to create. can we extract the logic to other functions and re use them? update should only call read

Reply:
Unsure about this one:

The only way to "update" this resource is to delete the old and create a new.  So this implementation seemed a good way.  But happy to change it.  Options I see are:

(a) As is -- call the canonical "delete" and "create" methods
(b) Refactor the common delete (and create) logic eg to a `doDelete` (and `doCreate`) which gets used by the canonical "delete" ("create") and "update"
(c) Just duplicate the delete and create logic here
(d) Mark the fields as `ForceNew: true` and make this method return an `UnexpectedInvocation` error

What do you recommend?

Reply:
(d) seems the right way:  as soon as all fields are `ForceNew` it complains if the `update` method even exists (nice)